lib/vminsertapi/server.go

```diff
@@ -0,0 +1,359 @@
+			bc, err := s.handshakeFunc(c, compressionLevel)
+			if err != nil {
+				if s.isStopping() {
+					// c is stopped inside VMInsertServer.MustStop
+					return
+				}
+				if handshake.IsTimeoutNetworkError(err) {
+					logger.Warnf("cannot complete vminsert handshake due to network timeout error with client %q: %s. "+
+						"If errors are transient and infrequent increase -rpc.handshakeTimeout and -vmstorageDialTimeout on client and server side. Check vminsert logs for errors", c.RemoteAddr(), err)
+				} else if handshake.IsClientNetworkError(err) {
+					logger.Warnf("cannot complete vminsert handshake due to network error with client %q: %s. "+
+						"Check vminsert logs for errors", c.RemoteAddr(), err)
+				} else if !handshake.IsTCPHealthcheck(err) {
+					logger.Errorf("cannot perform vminsert handshake with client %q: %s", c.RemoteAddr(), err)
+				}
+				_ = c.Close()
+				return
+			}
+			defer func() {
+				if !s.isStopping() {
+					logger.Infof("closing vminsert conn from %s", c.RemoteAddr())
+				}
+				_ = bc.Close()
+			}()
+
+			logger.Infof("processing vminsert conn from %s", c.RemoteAddr())
+			if err := s.processConn(bc); err != nil {
+				if s.isStopping() {
+					// c is stopped inside VMInsertServer.MustStop
+					return
+				}
+				s.vminsertConnErrors.Inc()
+				logger.Errorf("cannot process vminsert conn from %s: %s", c.RemoteAddr(), err)
+				return
+			}
+		}()
+	}
+}
+
+// MustStop gracefully stops s so it no longer touches s.storage after returning.
+func (s *VMInsertServer) MustStop() {
+	// Mark the server as stopping.
+	s.setIsStopping()
+
+	// Stop accepting new connections from vminsert.
+	if err := s.ln.Close(); err != nil {
+		logger.Panicf("FATAL: cannot close vminsert listener: %s", err)
+	}
+
+	// Close existing connections from vminsert, so the goroutines
+	// processing these connections are finished.
+	s.connsMap.CloseAll(s.connectionTimeout)
+
+	// Wait until all the goroutines processing vminsert conns are finished.
+	s.wg.Wait()
+}
+
+func (s *VMInsertServer) setIsStopping() {
+	s.stopFlag.Store(true)
+}
+
+func (s *VMInsertServer) isStopping() bool {
+	return s.stopFlag.Load()
+}
+
+func (s *VMInsertServer) processConn(bc *handshake.BufferedConn) error {
+	ctx := NewRequestCtx(bc)
+	for {
+		if err := s.processRequest(ctx); err != nil {
+			if errors.Is(err, io.EOF) {
+				// Remote client gracefully closed the connection.
+				return nil
+			}
+			return fmt.Errorf("cannot process vminsert request: %w", err)
+		}
+		if err := bc.Flush(); err != nil {
+			return fmt.Errorf("cannot flush compressed buffers: %w", err)
+		}
+	}
+}
+
+const maxRPCNameSize = 128
+
+func (s *VMInsertServer) processRequest(ctx *RequestCtx) error {
+	// Read rpcName
+	// Do not set deadline on reading rpcName, since it may take a
+	// lot of time for idle connection.
+	ctx.dataBuf = ctx.dataBuf[:0]
+	if ctx.bc.IsLegacy {
+		// fallback to prev API without RPC
+		return s.processWriteRowsLegacy(ctx)
```

Comment:
It would be good to log when we switch to legacy format. It might be helpful in future debug sessions.